rmw_fastrtps_cpp/src/rmw_wait.cpp

```diff
@@ -148,21 +148,28 @@ rmw_wait(
   // If wait_timeout is not null and either of its fields are nonzero, we have to wait
   bool hasToWait = (wait_timeout && (wait_timeout->sec > 0 || wait_timeout->nsec > 0)) ||
     !wait_timeout;
-  hasToWait &= !check_wait_set_for_data(subscriptions, guard_conditions, services, clients);
+  bool hasData = check_wait_set_for_data(subscriptions, guard_conditions, services, clients);
+  hasToWait &= !hasData;
 
   bool timeout = false;
-
-  if (hasToWait) {
-    if (!wait_timeout) {
+  if (hasToWait && wait_timeout) {
```

Comment:
I think this code will work fine. If I understand correctly in pseudo code it says

```
if (no data and timeout > 0)  // line 155
    wait until there is data with timeout
    return timeout error if still no data  // line 162
else
    if (no data and timeout == NULL)  // line 164
        wait until there is data
    if (no data and timeout == 0)  // line 171
        return timeout error
```

What do you think about separating having data initially from the timeout logic?

```
if (no data)
    if (timeout == NULL)
        wait until there is data
    else if (timeout > 0)
        wait until there is data with timeout
        return timeout error if still no data
    else
        return timeout error
```